File: TeamCode/src/main/cpp/sdk/sources/hardware/Modern_robotics_color_beacon.cpp
//
// Created by flemmi on 05.11.24.
//

#include "hardware/Modern_robotics_color_beacon.h"

namespace sdk {
    jclass Modern_robotics_color_beacon::jclazz;

    Modern_robotics_color_beacon::Modern_robotics_color_beacon(
            const jobject &modernRoboticsColorBeacon) : modernRoboticsColorBeacon(
            modernRoboticsColorBeacon) {}

    Modern_robotics_color_beacon::~Modern_robotics_color_beacon() {
        if (modernRoboticsColorBeacon) {
            attach_thread
            env->DeleteGlobalRef(modernRoboticsColorBeacon);
            modernRoboticsColorBeacon = nullptr;
        }
    }

    Modern_robotics_color_beacon &Modern_robotics_color_beacon::operator=(
            const Modern_robotics_color_beacon &modern_robotics_color_beacon) {
        if (this != &modern_robotics_color_beacon &&
            modern_robotics_color_beacon.modernRoboticsColorBeacon) {
            attach_thread
            this->modernRoboticsColorBeacon = env->NewGlobalRef(
                    modern_robotics_color_beacon.modernRoboticsColorBeacon);
        }
        return *this;
    }

    Modern_robotics_color_beacon &
    Modern_robotics_color_beacon::operator=(const jobject &modernRoboticsColorBeacon) {
        if (this->modernRoboticsColorBeacon) {
            attach_thread
            env->DeleteGlobalRef(this->modernRoboticsColorBeacon);
        }
        this->modernRoboticsColorBeacon = modernRoboticsColorBeacon;
        return *this;
    }

    void Modern_robotics_color_beacon::turn_LEDs_off() const {
        attach_thread
        env->CallVoidMethod(modernRoboticsColorBeacon,
                            env->GetMethodID(jclazz, "turnLEDsOff", "()V"));
    }

    void Modern_robotics_color_beacon::set_red() const {
        attach_thread
        env->CallVoidMethod(modernRoboticsColorBeacon, env->GetMethodID(jclazz, "setRed", "()V"));
    }

    void Modern_robotics_color_beacon::set_green() const {
        attach_thread
        env->CallVoidMethod(modernRoboticsColorBeacon, env->GetMethodID(jclazz, "setGreen", "()V"));
    }

    void Modern_robotics_color_beacon::set_yellow() const {
        attach_thread
        env->CallVoidMethod(modernRoboticsColorBeacon,
                            env->GetMethodID(jclazz, "setYellow", "()V"));
    }

    void Modern_robotics_color_beacon::set_blue() const {
        attach_thread
        env->CallVoidMethod(modernRoboticsColorBeacon, env->GetMethodID(jclazz, "setBlue", "()V"));
    }

    void Modern_robotics_color_beacon::set_purple() const {
        attach_thread
        env->CallVoidMethod(modernRoboticsColorBeacon,
                            env->GetMethodID(jclazz, "setPurple", "()V"));
    }

    void Modern_robotics_color_beacon::set_teal() const {
        attach_thread
        env->CallVoidMethod(modernRoboticsColorBeacon, env->GetMethodID(jclazz, "setTeal", "()V"));
    }

    void Modern_robotics_color_beacon::set_white() const {
        attach_thread
        env->CallVoidMethod(modernRoboticsColorBeacon, env->GetMethodID(jclazz, "setWhite", "()V"));
    }

    void Modern_robotics_color_beacon::set_color(const int &red, const int &green,
                                                 const int &blue) const {
        attach_thread
        env->CallVoidMethod(modernRoboticsColorBeacon,
                            env->GetMethodID(jclazz, "setColor", "(III)V"), static_cast<jint>(red),
                            static_cast<jint>(green), static_cast<jint>(blue));
    }

    bool Modern_robotics_color_beacon::do_initialize() const {
        attach_thread
        return env->CallBooleanMethod(modernRoboticsColorBeacon,
                                      env->GetMethodID(jclazz, "doInitialize", "()Z"));
    }

    std::string Modern_robotics_color_beacon::get_device_name() const {
        attach_thread
        auto jstr = (jstring) env->CallObjectMethod(modernRoboticsColorBeacon,
                                                    env->GetMethodID(jclazz, "getDeviceName",
                                                                     "()Ljava/lang/String;"));

        const char *cstr = env->GetStringUTFChars(jstr, nullptr);
        std::string str(cstr);
        env->DeleteLocalRef(jstr);

        return str;
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/sdk/hardware/ModernRoboticsColorBeacon.java
package org.firstinspires.ftc.teamcode.sdk.hardware;

import com.qualcomm.robotcore.hardware.I2cAddr;
import com.qualcomm.robotcore.hardware.I2cDevice;
import com.qualcomm.robotcore.hardware.I2cDeviceSynch;
import com.qualcomm.robotcore.hardware.I2cDeviceSynchDevice;
import com.qualcomm.robotcore.hardware.configuration.annotations.DeviceProperties;
import com.qualcomm.robotcore.hardware.configuration.annotations.I2cDeviceType;

@SuppressWarnings("unused")
@I2cDeviceType()
@DeviceProperties(name = "Modern Robotics Color Beacon", description = "Color Beacon", xmlTag = "ColorBeacon")
public class ModernRoboticsColorBeacon extends I2cDeviceSynchDevice<I2cDeviceSynch> {

    I2cDevice colorBeacon;
    I2cDeviceSynch colorBeaconSynch;

    public ModernRoboticsColorBeacon(I2cDeviceSynch colorBeaconSynch) {
        //default address: 0x4c
        super(colorBeaconSynch, true);
        this.colorBeaconSynch = colorBeaconSynch;
        this.colorBeaconSynch.setI2cAddress(I2cAddr.create8bit(0x4c));
        super.registerArmingStateCallback(false);
        this.colorBeaconSynch.engage();
        setColor(0, 0, 0);
    }

    public void turnLEDsOff() {
        colorBeaconSynch.write8(4, 0); // off
    }

    public void setRed() {
        colorBeaconSynch.write8(4, 1); // red
    }

    public void setGreen() {
        colorBeaconSynch.write8(4, 2); // green
    }

    public void setYellow() {
        colorBeaconSynch.write8(4, 3); // yellow
    }

    public void setBlue() {
        colorBeaconSynch.write8(4, 4); // blue
    }

    public void setPurple() {
        colorBeaconSynch.write8(4, 5); // purple
    }

    public void setTeal() {
        colorBeaconSynch.write8(4, 6); // teal
    }

    public void setWhite() {
        colorBeaconSynch.write8(4, 7); // white
    }

    public void setColor(int red, int green, int blue) {
        colorBeaconSynch.write8(4, 8); // custom color mode
        colorBeaconSynch.write8(5, red); // red
        colorBeaconSynch.write8(6, green); // green
        colorBeaconSynch.write8(7, blue); // blue
    }

    @Override
    protected boolean doInitialize() {
        return true;
    }

    @Override
    public Manufacturer getManufacturer() {
        return Manufacturer.ModernRobotics;
    }

    @Override
    public String getDeviceName() {
        return "ColorBeacon";
    }
}
